lib/rubocop/cop/rspec/message_spies.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that message expectations are set using spies.
+      #
+      # This cop can be configured in your configuration using the
+      # `EnforcedStyle` option and supports `--auto-gen-config`.
+      #
+      # @example `EnforcedStyle: have_received`
+      #
+      #   # bad
+      #   expect(foo).to receive(:bar)
+      #
+      #   # good
+      #   expect(foo).to have_received(:bar)
+      #
+      # @example `EnforcedStyle: receive`
+      #
+      #   # bad
+      #   expect(foo).to have_received(:bar)
+      #
+      #   # good
+      #   expect(foo).to receive(:bar)
+      #
+      class MessageSpies < Cop
+        include RuboCop::RSpec::SpecOnly, ConfigurableEnforcedStyle
+
+        MSG_RECEIVE = 'Prefer `receive` for setting message ' \
+          'expectations.'.freeze
+        MSG_HAVE_RECEIVED = 'Prefer `have_received` for setting message ' \
+          'expectations. You will need to set up `%s` as a spy, using either ' \
+          '`allow` or `spy`.'.freeze
+
+        SUPPORTED_STYLES = %w(have_received receive).freeze
+
+        def_node_matcher :message_expectation, %(
+          (send (send nil :expect (send nil $_)) :to ...)
+        )
+
+        def_node_search :receive_message, %(
+          $(send nil {:receive :have_received} ...)
+        )
+
+        def on_send(node)
+          message_expectation(node) do |receiver|
```

Comment:
I usually try to avoid multiple levels of block nesting like this (happy to elaborate on why if you are curious). I usually opt to create a private method which more clearly conveys what nodes I'm filtering down to. In this case I might do something like this:

```ruby
def on_send(node)
  receive_message_matcher(node) do |receiver, message_matcher|
    return correct_style_detected if preferred_style?(message_matcher)

    add_offense(message_matcher, :selector, error_message(receiver)) do
      opposite_style_detected
    end
  end
end

private

def receive_message_matcher(node)
  return unless (receiver = message_expectation(node))

  receive_message(node) { |match| yield(receiver, match) }
end
```